File: c/problem204.c
#include "leetcode.h"

int countPrimes(int n) {
    if (n < 2) return 0;
    int* primes = (int*)calloc(n, sizeof(int));
    primes[0] = 1;
    primes[1] = 1;
    int x = sqrt(n);
    for(int i = 2; i < x; ++i)
    {
        if(primes[i] == 0)
        {
            for(int j = i * i; j < n; j += i)
            {
                primes[j] = 1;
            }
        }
    }
    int c = 0;
    for(int i = 10000000; i < n; ++i)
    {
        if(primes[i] == 0)
        {
            ++c;
            printf("%d\n", i);
        }
    }
    free(primes);
    return c;
}

void test()
{
    int n = 100;
    //printf("%d: %d\n", n, countPrimes(n));
    //n = 10000000;
    //printf("%d: %d\n", n, countPrimes(n));
    n = 100000000;
    printf("%d: %d\n", n, countPrimes(n));
}

File: python/problem204.py
import math
def countPrimes(n):
	if n <= 2:
		return 0
	primes = [True] * n
	primes[0] = False
	primes[1] = False
	x = int(math.sqrt(n))
	for i in xrange(x):
		if primes[i]:
			for j in range(i * i, n, i):
				primes[j] = False
	c = 0
	for i in range(n):
		if primes[i]:
			c += 1
	return c

def countPrimes2(n):
	'''
	Runs in linear time
	'''
	if n <= 2:
		return []
	primes = [True] * n
	primes[0] = False
	primes[1] = False
	results = [0] * (n / 2)
	i = 2
	k = 0
	while i < n:
		if primes[i]:
			results[k] = i
			k += 1
		j = 0
		while j < k:
			p = results[j]
			if i * p >= n:
				break
			primes[i * p] = False
			if i % p == 0:
				break
			j += 1
		i += 1
	return k



def test():
	n = 10000000
	print countPrimes2(n)
	print countPrimes(n)

test()
